index.js

```diff
@@ -180,2669 +180,2588 @@
+  );
+};
+
+const NullaryTypeWithUrl = name => supertypes => test => (
+  _NullaryType (name) (functionUrl (name)) (supertypes) (test)
+);
+
+const UnaryTypeWithUrl = name => supertypes => test => _1 => (
+  def (name)
+      ({})
+      ([Type, Type])
+      (_UnaryType (name) (functionUrl (name)) (supertypes) (test) (_1))
+);
+
+const BinaryTypeWithUrl = name => supertypes => test => _1 => _2 => (
+  def (name)
+      ({})
+      ([Type, Type, Type])
+      (_BinaryType (name) (functionUrl (name)) (supertypes) (test) (_1) (_2))
+);
+
+const mkdef = opts => name => constraints => types => impl => {
+  if (!opts.checkTypes) /* c8 ignore next */ return impl;
+  const typeInfo = {
+    name,
+    constraints,
+    types: types.length === 1 ? [NoArguments, ...types] : types,
   };
+  return withTypeChecking (opts.env, typeInfo, impl);
+};
 
-  //  _Type :: ... -> Type
-  const _Type = (
-    type,       // :: String
-    name,       // :: String
-    url,        // :: String
-    arity,      // :: NonNegativeInteger
-    format,
-    // :: Nullable ((String -> String, String -> String -> String) -> String)
-    supertypes, // :: Array Type
-    test,       // :: Array Type -> Any -> Boolean
-    tuples      // :: Array (Array3 String (a -> Array b) Type)
-  ) => globalThis.Object.assign (
-    globalThis.Object.create (Type$prototype, {
-      _extractors: {
-        value: tuples.reduce ((extractors, [k, e]) => ((
-          extractors[k] = e,
-          extractors
-        )), {}),
-      },
-      _test: {
-        value: test,
-      },
-      extractors: {
-        value: tuples.reduce ((extractors, [k, e]) => ((
-          extractors[k] = x => toArray (e (x)),
-          extractors
-        )), {}),
-      },
-      format: {
-        value: format || ((outer, inner) =>
-          outer (name) +
-          Z.foldMap (
-            globalThis.String,
-            ([k, , t]) => (
-              t.arity > 0
-              ? outer (' ') + outer ('(') + inner (k) (show (t)) + outer (')')
-              : outer (' ')               + inner (k) (show (t))
-            ),
-            tuples
-          )
-        ),
-      },
-    }),
-    {
-      arity,  // number of type parameters
-      keys: tuples.map (([k]) => k),
-      name,
-      supertypes,
-      type,
-      types: tuples.reduce ((types, [k, , t]) => ((types[k] = t, types)), {}),
-      url,
-    }
-  );
+const production = globalThis.process?.env?.NODE_ENV === 'production';
 
-  const BINARY        = 'BINARY';
-  const FUNCTION      = 'FUNCTION';
-  const INCONSISTENT  = 'INCONSISTENT';
-  const NO_ARGUMENTS  = 'NO_ARGUMENTS';
-  const NULLARY       = 'NULLARY';
-  const RECORD        = 'RECORD';
-  const UNARY         = 'UNARY';
-  const UNKNOWN       = 'UNKNOWN';
-  const VARIABLE      = 'VARIABLE';
-
-  //  Inconsistent :: Type
-  const Inconsistent = _Type (
-    INCONSISTENT,
-    '',
-    '',
-    0,
-    (outer, inner) => '???',
-    [],
-    null,
-    []
-  );
+const env = [];
```

Comment:
`env` is now defined here, provided to `mkdef` below, and populated later. :see_no_evil: